src/envoy/http/service_control/client_cache.cc

```diff
@@ -381,37 +388,50 @@ void ClientCache::handleCheckResponse(const Status& http_status,
     }
   } else {
     if (translate_non_5xx) {
+      // Most likely an auth error in ESPv2 or API producer deployment.
+      filter_stats_.filter_.denied_producer_error_.inc();
+
       // This is not caused by a client request error, so translate
       // non-5xx error codes to 500 Internal Server Error. Error message
       // contains details on the original error (including the original
       // HTTP status code).
       Status scrubbed_status(Code::INTERNAL, final_status.error_message());
       on_done(scrubbed_status, response_info);
     } else {
+      // Stats already incremented for this case.
       on_done(final_status, response_info);
     }
   }
   delete response;
 }
 
-void ClientCache::callQuota(
-    const ::google::api::servicecontrol::v1::AllocateQuotaRequest& request,
-    std::function<void(const ::google::protobuf::util::Status& status)>
-        on_done) {
+void ClientCache::callQuota(const AllocateQuotaRequest& request,
+                            QuotaDoneFunc on_done) {
   auto* response = new AllocateQuotaResponse;
-  client_->Quota(
-      request, response, [this, response, on_done](const Status& status) {
-        if (status.ok()) {
-          on_done(::espv2::api_proxy::service_control::RequestBuilder::
-                      ConvertAllocateQuotaResponse(*response,
-                                                   config_.service_name()));
-        } else {
-          on_done(Status(static_cast<google::protobuf::util::error::Code>(
-                             status.error_code()),
-                         status.error_message()));
-        }
-        delete response;
-      });
+  client_->Quota(request, response,
+                 [this, response, on_done](const Status& status) {
+                   handleQuotaResponse(status, response, on_done);
+                 });
+}
+
+void ClientCache::handleQuotaResponse(const Status& http_status,
```

Comment:
we should call it handleQuotaOnDone.   This is called from or after cache.  It is different with QuotaResponse which I feel it is from quota_transport which is the actual response from the server.   Regardless, please add this comment.

e.g. response proto message it is from the cache,  except the first cache miss,  which is an empty one.

Reply:
Done.